dan/webserver/webserver.py

```diff
@@ -68,42 +66,34 @@ async def chat(request, ws):
                 }
                 
                 # store message
-                messages.append(message)
-                if len(messages) > 20:  # keep only last 20 messages
-                    messages.pop(0)
+                __message_append(message)
+                
                 await _send_to_all_clients(message)
             
             elif data['type'] == 'pin_message':
-                #create pinned message object
                 pinned_message = {
                     'type': 'pinned_message',
-                    'username': username,
+                    'username': data['username'],
                     'text': data['text']
                 }
-                #stores all pinned messages
-                pinned_messages.append(pinned_message)
-                #sends pinned messages to every user
-                await _send_to_all_clients(pinned_message)
+                if pinned_messages not in pinned_messages:
+                    __pinned_message_append(pinned_message)
+                    # sends pinned messages to every user
+                    await _send_to_all_clients(pinned_message)
             
             elif data['type'] == 'unpin_message':
                 remove_text = data['text']
-                leftover_pinned = []
```

Comment:
just removed messages from list instead of creating a copy. Less memory used this way.